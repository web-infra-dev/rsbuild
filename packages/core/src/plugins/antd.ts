import { getAntdMajorVersion } from '@modern-js/utils';
import type { RsbuildTarget, DefaultRsbuildPlugin } from '@rsbuild/shared';

export const pluginAntd = (): DefaultRsbuildPlugin => ({
  name: `plugin-antd`,

  setup(api) {
    api.modifyRsbuildConfig((builderConfig) => {
      builderConfig.source ??= {};

      if (
        builderConfig.source.transformImport === false ||
        builderConfig.source.transformImport?.some(
          (item) => item.libraryName === 'antd',
        )
      ) {
        return;
      }

      const antdMajorVersion = getAntdMajorVersion(api.context.rootPath);
      // antd >= v5 no longer need babel-plugin-import
      // see: https://ant.design/docs/react/migration-v5#remove-babel-plugin-import
      if (antdMajorVersion && antdMajorVersion < 5) {
        builderConfig.source ??= {};
        builderConfig.source.transformImport = [
          ...(builderConfig.source.transformImport || []),
          {
            libraryName: 'antd',
            libraryDirectory: useSSR(api.context.target) ? 'lib' : 'es',
            style: true,
          },
        ];
      }
    });
  },
});

export function useSSR(target: RsbuildTarget | RsbuildTarget[]) {
  return (Array.isArray(target) ? target : [target]).some((item) =>
    ['node', 'service-worker'].includes(item),
  );
}
