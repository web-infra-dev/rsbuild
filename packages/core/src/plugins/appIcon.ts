import path from 'node:path';
import { promisify } from 'node:util';
import { lookup } from 'mrmime';
import {
  ensureAssetPrefix,
  fileExistsByCompilation,
  getPublicPathFromCompiler,
  isURL,
} from '../helpers';
import type { AppIconItem, HtmlBasicTag, RsbuildPlugin } from '../types';

type IconExtra = {
  sizes: string;
  isURL?: boolean;
  absolutePath: string;
  relativePath: string;
  requestPath: string;
  mimeType?: string;
};

export const pluginAppIcon = (): RsbuildPlugin => ({
  name: 'rsbuild:app-icon',

  setup(api) {
    const htmlTagsMap = new Map<string, HtmlBasicTag[]>();
    const iconFormatMap = new Map<string, IconExtra>();

    const formatIcon = (
      icon: AppIconItem,
      distDir: string,
      publicPath: string,
    ): AppIconItem & IconExtra => {
      const { src, size } = icon;
      const cached = iconFormatMap.get(src);

      if (cached) {
        return { ...cached, ...icon };
      }

      const sizes = `${size}x${size}`;

      if (isURL(src)) {
        const paths = {
          sizes,
          isURL: true,
          requestPath: src,
          absolutePath: src,
          relativePath: src,
          mimeType: lookup(src),
        };

        iconFormatMap.set(src, paths);

        return { ...paths, ...icon };
      }

      const absolutePath = path.isAbsolute(src)
        ? src
        : path.join(api.context.rootPath, src);
      const relativePath = path.posix.join(
        distDir,
        path.basename(absolutePath),
      );
      const requestPath = ensureAssetPrefix(relativePath, publicPath);

      const paths = {
        sizes,
        requestPath,
        absolutePath,
        relativePath,
        mimeType: lookup(absolutePath),
      };

      iconFormatMap.set(src, paths);

      return { ...paths, ...icon };
    };

    api.processAssets(
      { stage: 'additional' },
      async ({ compilation, environment, sources }) => {
        const { config } = environment;
        const { appIcon } = config.html;

        if (!appIcon) {
          return;
        }

        const distDir = config.output.distPath.image;
        const manifestFile = appIcon.filename ?? 'manifest.webmanifest';
        const publicPath = getPublicPathFromCompiler(compilation);
        const icons = appIcon.icons.map((icon) =>
          formatIcon(icon, distDir, publicPath),
        );
        const tags: HtmlBasicTag[] = [];

        for (const icon of icons) {
          if (icon.target === 'web-app-manifest' && !appIcon.name) {
            throw new Error(
              "[rsbuild:app-icon] `appIcon.name` is required when `target` is 'web-app-manifest'.",
            );
          }

          if (!icon.isURL) {
            if (!compilation.inputFileSystem) {
              throw new Error(
                `[rsbuild:app-icon] 'compilation.inputFileSystem' is not available.`,
              );
            }

            if (
              !(await fileExistsByCompilation(compilation, icon.absolutePath))
            ) {
              throw new Error(
                `[rsbuild:app-icon] Can not find the app icon, please check if the '${icon.relativePath}' file exists'.`,
              );
            }

            const source = await promisify(
              compilation.inputFileSystem.readFile,
            )(icon.absolutePath);

            if (!source) {
              throw new Error(
                `[rsbuild:app-icon] Failed to read the app icon file, please check if the '${icon.relativePath}' file exists'.`,
              );
            }

            compilation.emitAsset(
              icon.relativePath,
              new sources.RawSource(source),
            );
          }

          if (
            icon.target === 'apple-touch-icon' ||
            (!icon.target && icon.size < 200)
          ) {
            tags.push({
              tag: 'link',
              attrs: {
                rel: 'apple-touch-icon',
                sizes: icon.sizes,
                href: icon.requestPath,
              },
            });
          }
        }

        if (appIcon.name) {
          const manifestIcons = icons
            .filter(
              (icon) => icon.target === 'web-app-manifest' || !icon.target,
            )
            .map((icon) => {
              const result = {
                src: icon.requestPath,
                sizes: icon.sizes,
              };
              if (icon.mimeType) {
                return { ...result, type: icon.mimeType };
              }
              return result;
            });

          const manifest = {
            name: appIcon.name,
            icons: manifestIcons,
          };

          compilation.emitAsset(
            manifestFile,
            new sources.RawSource(JSON.stringify(manifest)),
          );

          tags.push({
            tag: 'link',
            attrs: {
              rel: 'manifest',
              href: ensureAssetPrefix(manifestFile, publicPath),
            },
          });
        }

        if (tags.length) {
          htmlTagsMap.set(environment.name, tags);
        }
      },
    );

    api.modifyHTMLTags(({ headTags, bodyTags }, { environment }) => {
      const tags = htmlTagsMap.get(environment.name);
      if (tags) {
        headTags.unshift(...tags);
      }
      return { headTags, bodyTags };
    });

    api.onCloseDevServer(() => {
      htmlTagsMap.clear();
      iconFormatMap.clear();
    });
  },
});
