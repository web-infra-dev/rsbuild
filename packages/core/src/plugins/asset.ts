import path from 'path';
import {
  getDistPath,
  getFilename,
  chainStaticAssetRule,
} from '@rsbuild/shared';
import type { DefaultRsbuildPlugin } from '@rsbuild/shared';

export function getRegExpForExts(exts: string[]): RegExp {
  const matcher = exts
    .map((ext) => ext.trim())
    .map((ext) => (ext.startsWith('.') ? ext.slice(1) : ext))
    .join('|');

  return new RegExp(
    exts.length === 1 ? `\\.${matcher}$` : `\\.(${matcher})$`,
    'i',
  );
}

export const pluginAsset = (
  assetType: 'image' | 'media' | 'font' | 'svg',
  exts: string[],
): DefaultRsbuildPlugin => ({
  name: `plugin-${assetType}`,

  setup(api) {
    api.modifyBundlerChain((chain, { isProd }) => {
      const config = api.getNormalizedConfig();
      const regExp = getRegExpForExts(exts);
      const distDir = getDistPath(config.output, assetType);
      const filename = getFilename(config.output, assetType, isProd);

      const maxSize = config.output.dataUriLimit[assetType];

      const rule = chain.module.rule(assetType).test(regExp);

      chainStaticAssetRule({
        rule,
        maxSize,
        filename: path.posix.join(distDir, filename),
        assetType,
      });
    });
  },
});
