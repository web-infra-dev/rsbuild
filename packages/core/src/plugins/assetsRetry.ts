import {
  getDistPath,
  isHtmlDisabled,
  DefaultRsbuildPlugin,
} from '@rsbuild/shared';

export function pluginAssetsRetry(): DefaultRsbuildPlugin {
  return {
    name: 'plugin-assets-retry',
    setup(api) {
      api.modifyBundlerChain(
        async (chain, { CHAIN_ID, target, HtmlPlugin }) => {
          const config = api.getNormalizedConfig();

          if (!config.output.assetsRetry || isHtmlDisabled(config, target)) {
            return;
          }

          const { AssetsRetryPlugin } = await import('@rsbuild/shared');
          const distDir = getDistPath(config.output, 'js');

          const { assetsRetry = {} } = config.output;

          // assetsRetry.crossOrigin should be same as html.crossorigin by default
          if (assetsRetry.crossOrigin === undefined) {
            assetsRetry.crossOrigin = config.html.crossorigin;
          }

          chain.plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY).use(AssetsRetryPlugin, [
            {
              ...assetsRetry,
              distDir,
              HtmlPlugin,
            },
          ]);
        },
      );
    },
  };
}
