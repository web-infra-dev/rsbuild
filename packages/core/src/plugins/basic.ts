import path from 'node:path';
import { TARGET_ID_MAP, getJsSourceMap, isUsingHMR } from '@rsbuild/shared';
import type { RsbuildPlugin } from '../types';

/**
 * Provide some basic configs of rspack
 */
export const pluginBasic = (): RsbuildPlugin => ({
  name: 'rsbuild:basic',

  setup(api) {
    api.modifyBundlerChain(
      (chain, { env, isProd, target, bundler, CHAIN_ID }) => {
        const config = api.getNormalizedConfig();

        chain.name(TARGET_ID_MAP[target]);

        chain.devtool(getJsSourceMap(config));

        // The base directory for resolving entry points and loaders from the configuration.
        chain.context(api.context.rootPath);

        chain.mode(isProd ? 'production' : 'development');

        chain.merge({
          infrastructureLogging: {
            // Using `error` level to avoid `cache.PackFileCacheStrategy` logs
            level: 'error',
          },
        });

        // Disable performance hints, these logs are too complex
        chain.performance.hints(false);

        // Align with the futureDefaults of webpack 6
        chain.module.parser.merge({
          javascript: {
            exportsPresence: 'error',
          },
        });

        const usingHMR = isUsingHMR(config, { target, isProd });

        if (usingHMR) {
          chain
            .plugin(CHAIN_ID.PLUGIN.HMR)
            .use(bundler.HotModuleReplacementPlugin);
        }

        if (env === 'development') {
          // Set correct path for source map
          // this helps VS Code break points working correctly in monorepo
          chain.output.devtoolModuleFilenameTemplate(
            (info: { absoluteResourcePath: string }) =>
              path.resolve(info.absoluteResourcePath).replace(/\\/g, '/'),
          );
        }
      },
    );
  },
});
