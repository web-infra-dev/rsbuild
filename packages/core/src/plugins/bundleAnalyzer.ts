import { isProd } from '../helpers';
import type {
  NormalizedEnvironmentConfig,
  RsbuildConfig,
  RsbuildPlugin,
} from '../types';

// There are two ways to enable the bundle analyzer:
// 1. Set environment variable `BUNDLE_ANALYZE`
// 2. Set performance.bundleAnalyze config
const isUseAnalyzer = (config: RsbuildConfig | NormalizedEnvironmentConfig) =>
  process.env.BUNDLE_ANALYZE || config.performance?.bundleAnalyze;

export function pluginBundleAnalyzer(): RsbuildPlugin {
  return {
    name: 'rsbuild:bundle-analyzer',

    setup(api) {
      api.modifyRsbuildConfig({
        order: 'post',
        handler: (config) => {
          if (isProd()) {
            return;
          }

          const useAnalyzer =
            isUseAnalyzer(config) ||
            Object.values(config.environments || []).some((config) =>
              isUseAnalyzer(config),
            );

          if (!useAnalyzer) {
            return;
          }

          // webpack-bundle-analyze needs to read assets from disk
          config.dev ||= {};
          config.dev.writeToDisk = true;

          return config;
        },
      });

      api.modifyBundlerChain(async (chain, { CHAIN_ID, environment }) => {
        const config = api.getNormalizedConfig({ environment });

        if (!isUseAnalyzer(config)) {
          return;
        }

        const { default: BundleAnalyzer } = await import(
          '@rsbuild/shared/webpack-bundle-analyzer'
        );

        chain
          .plugin(CHAIN_ID.PLUGIN.BUNDLE_ANALYZER)
          .use(BundleAnalyzer.BundleAnalyzerPlugin, [
            {
              analyzerMode: 'static',
              openAnalyzer: false,
              reportFilename: `report-${environment}.html`,
              ...(config.performance.bundleAnalyze || {}),
            },
          ]);
      });
    },
  };
}
