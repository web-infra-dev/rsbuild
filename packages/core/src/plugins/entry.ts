import { castArray, color } from '@rsbuild/shared';
import type { EntryDescription } from '@rspack/core';
import { createVirtualModule } from '../helpers';
import type { RsbuildPlugin } from '../types';

export const pluginEntry = (): RsbuildPlugin => ({
  name: 'rsbuild:entry',

  setup(api) {
    api.modifyBundlerChain(async (chain, { environment, isServer }) => {
      const config = api.getNormalizedConfig({ environment });
      const { preEntry } = config.source;
      const { entry } = api.context.environments[environment];

      const injectCoreJsEntry = config.output.polyfill === 'entry' && !isServer;

      for (const entryName of Object.keys(entry)) {
        const entryPoint = chain.entry(entryName);
        const addEntry = (item: string | EntryDescription) => {
          entryPoint.add(item);
        };

        preEntry.forEach(addEntry);

        if (injectCoreJsEntry) {
          addEntry(createVirtualModule('import "core-js";'));
        }

        castArray(entry[entryName]).forEach(addEntry);
      }
    });

    api.onBeforeCreateCompiler(({ bundlerConfigs }) => {
      if (bundlerConfigs.every((config) => !config.entry)) {
        throw new Error(
          `Could not find any entry module, please make sure that ${color.cyan(
            'src/index.(ts|js|tsx|jsx|mjs|cjs)',
          )} exists, or customize entry through the ${color.cyan(
            'source.entry',
          )} configuration.`,
        );
      }
    });
  },
});
