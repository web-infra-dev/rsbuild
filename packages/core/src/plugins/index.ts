import {
  MEDIA_EXTENSIONS,
  FONT_EXTENSIONS,
  IMAGE_EXTENSIONS,
  Plugins,
} from '@rsbuild/shared';

export const plugins: Plugins = {
  html: () => import('./html').then((m) => m.pluginHtml()),
  cleanOutput: () => import('./cleanOutput').then((m) => m.pluginCleanOutput()),
  startUrl: () => import('./startUrl').then((m) => m.pluginStartUrl()),
  fileSize: () => import('./fileSize').then((m) => m.pluginFileSize()),
  devtool: () => import('./devtool').then((m) => m.pluginDevtool()),
  target: () => import('./target').then((m) => m.pluginTarget()),
  entry: () => import('./entry').then((m) => m.pluginEntry()),
  cache: () => import('./cache').then((m) => m.pluginCache()),
  yaml: () => import('./yaml').then((m) => m.pluginYaml()),
  toml: () => import('./toml').then((m) => m.pluginToml()),
  svg: () => import('./svg').then((m) => m.pluginSvg()),
  splitChunks: () => import('./splitChunks').then((m) => m.pluginSplitChunks()),
  inlineChunk: () => import('./inlineChunk').then((m) => m.pluginInlineChunk()),
  bundleAnalyzer: () =>
    import('./bundleAnalyzer').then((m) => m.pluginBundleAnalyzer()),
  font: () =>
    import('./asset').then((m) => m.pluginAsset('font', FONT_EXTENSIONS)),
  image: () =>
    import('./asset').then((m) => m.pluginAsset('image', IMAGE_EXTENSIONS)),
  media: () =>
    import('./asset').then((m) => m.pluginAsset('media', MEDIA_EXTENSIONS)),
  assetsRetry: () => import('./assetsRetry').then((m) => m.pluginAssetsRetry()),
  antd: () => import('./antd').then((m) => m.pluginAntd()),
  arco: () => import('./arco').then((m) => m.pluginArco()),
  checkSyntax: () => import('./checkSyntax').then((m) => m.pluginCheckSyntax()),
  rem: () => import('./rem').then((m) => m.pluginRem()),
  wasm: () => import('./wasm').then((m) => m.pluginWasm()),
  moment: () => import('./moment').then((m) => m.pluginMoment()),
  nodeAddons: () => import('./nodeAddons').then((m) => m.pluginNodeAddons()),
  externals: () => import('./externals').then((m) => m.pluginExternals()),
  networkPerformance: () =>
    import('./networkPerformance').then((m) => m.pluginNetworkPerformance()),
  preloadOrPrefetch: () =>
    import('./preloadOrPrefetch').then((m) => m.pluginPreloadOrPrefetch()),
  performance: () => import('./performance').then((m) => m.pluginPerformance()),
  define: () => import('./define').then((m) => m.pluginDefine()),
};
