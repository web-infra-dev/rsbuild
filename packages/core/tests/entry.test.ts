import { createStubRsbuild } from '@scripts/test-helper';
import { pluginEntry } from '../src/plugins/entry';

describe('plugin-entry', () => {
  const cases = [
    {
      name: 'should set entry correctly',
      entry: {
        main: './src/main.ts',
      },
      preEntry: [],
      expected: {
        entry: {
          main: ['./src/main.ts'],
        },
      },
    },
    {
      name: 'should set multiple entry correctly',
      entry: {
        foo: ['./src/polyfill.ts', './src/foo.ts'],
        bar: ['./src/polyfill.ts', './src/bar.ts'],
      },
      preEntry: [],
      expected: {
        entry: {
          bar: ['./src/polyfill.ts', './src/bar.ts'],
          foo: ['./src/polyfill.ts', './src/foo.ts'],
        },
      },
    },
    {
      name: 'should set pre-entry correctly',
      entry: {
        foo: ['./src/polyfill.ts', './src/foo.ts'],
        bar: './src/bar.ts',
      },
      preEntry: ['./src/pre-entry.ts'],
      expected: {
        entry: {
          bar: ['./src/pre-entry.ts', './src/bar.ts'],
          foo: ['./src/pre-entry.ts', './src/polyfill.ts', './src/foo.ts'],
        },
      },
    },
  ];

  it.each(cases)('$name', async (item) => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEntry()],
      rsbuildConfig: {
        source: {
          entry: item.entry as unknown as Record<string, string | string[]>,
          preEntry: item.preEntry,
        },
      },
    });

    const config = await rsbuild.unwrapConfig();

    expect(config).toEqual(item.expected);
  });

  it('should apply environments entry config correctly', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEntry()],
      rsbuildConfig: {
        environments: {
          web: {
            source: {
              entry: {
                index: './src/index.client',
              },
            },
          },
          ssr: {
            source: {
              entry: {
                index: './src/index.ssr',
              },
            },
          },
        },
      },
    });

    const configs = await rsbuild.initConfigs();
    expect(configs).toMatchSnapshot();
  });

  it('should apply different environments entry config correctly', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEntry()],
      rsbuildConfig: {
        environments: {
          web: {
            source: {
              entry: {
                index: './src/index.client',
              },
            },
          },
          ssr: {
            source: {
              entry: {
                main: './src/index.ssr',
              },
            },
          },
        },
      },
    });

    const configs = await rsbuild.initConfigs();

    expect(configs).toMatchSnapshot();
  });
});
