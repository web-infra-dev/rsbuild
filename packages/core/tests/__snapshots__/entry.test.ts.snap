// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-entry > should apply different environments entry config correctly 1`] = `
[
  {
    "entry": {
      "index": [
        "src/index.client",
      ],
    },
  },
  {
    "entry": {
      "main": [
        "src/index.ssr",
      ],
    },
  },
]
`;

exports[`plugin-entry > should apply environments entry config correctly 1`] = `
[
  {
    "entry": {
      "index": [
        "src/index.client",
      ],
    },
  },
  {
    "entry": {
      "index": [
        "src/index.ssr",
      ],
    },
  },
]
`;
